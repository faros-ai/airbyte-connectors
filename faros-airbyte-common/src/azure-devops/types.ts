import {AxiosInstance} from 'axios';
import {IBuildApi} from 'azure-devops-node-api/BuildApi';
import {ICoreApi} from 'azure-devops-node-api/CoreApi';
import {IGitApi} from 'azure-devops-node-api/GitApi';
import {
  Build as BaseBuild,
  BuildArtifact,
  TimelineRecord as BaseTimelineRecord,
} from 'azure-devops-node-api/interfaces/BuildInterfaces';
import {IdentityRef} from 'azure-devops-node-api/interfaces/common/VSSInterfaces';
import * as GitInterfaces from 'azure-devops-node-api/interfaces/GitInterfaces';
import {GraphUser} from 'azure-devops-node-api/interfaces/GraphInterfaces';
import {Pipeline as BasePipeline} from 'azure-devops-node-api/interfaces/PipelinesInterfaces';
import {ProjectReference} from 'azure-devops-node-api/interfaces/ReleaseInterfaces';
import {CodeCoverageStatistics} from 'azure-devops-node-api/interfaces/TestInterfaces';
import {IPipelinesApi} from 'azure-devops-node-api/PipelinesApi';
import {IReleaseApi} from 'azure-devops-node-api/ReleaseApi';
import {ITestApi} from 'azure-devops-node-api/TestApi';
import {IWorkItemTrackingApi} from 'azure-devops-node-api/WorkItemTrackingApi';

export type DevOpsCloud = {
  type: 'cloud';
};
export type DevOpsServer = {
  type: 'server';
  api_url: string;
};
export type AzureDevOpsInstance = DevOpsCloud | DevOpsServer;

export interface AzureDevOpsConfig {
  readonly instance?: AzureDevOpsInstance;
  readonly access_token: string;
  readonly organization: string;
  readonly projects?: ReadonlyArray<string>;
  readonly cutoff_days?: number;
  readonly page_size?: number;
  readonly max_retries?: number;
  readonly request_timeout?: number;
}

export interface AzureDevOpsClient {
  readonly build: IBuildApi;
  readonly core: ICoreApi;
  readonly git: IGitApi;
  readonly wit: IWorkItemTrackingApi;
  readonly pipelines: IPipelinesApi;
  readonly release: IReleaseApi;
  readonly test: ITestApi;
  readonly graph?: AxiosInstance;
}

export type User = GraphUser | IdentityRef;

export interface GraphUserResponse {
  count: number;
  value: GraphUser[];
}

export interface Pipeline extends BasePipeline {
  project: ProjectReference;
}
// Ensure Build reason, status, and result enums are strings
export interface Build extends Omit<BaseBuild, 'reason' | 'status' | 'result'> {
  artifacts: BuildArtifact[];
  coverageStats: CodeCoverageStatistics[];
  jobs: TimelineRecord[];
  reason: string;
  status: string;
  result: string;
}

export interface TimelineRecord
  extends Omit<BaseTimelineRecord, 'result' | 'state'> {
  result: string;
  state: string;
}

export interface Tag extends GitInterfaces.GitRef {
  commit?: GitInterfaces.GitAnnotatedTag;
}

export interface Repository extends GitInterfaces.GitRepository {
  branches?: GitInterfaces.GitBranchStats[];
  tags?: Tag[];
}

export interface PullRequest
  extends Omit<GitInterfaces.GitPullRequest, 'status' | 'mergeStatus'> {
  status: string;
  mergeStatus: string;
  threads: GitInterfaces.GitPullRequestCommentThread[];
}

export interface Commit extends GitInterfaces.GitCommitRef {
  repository?: Repository;
  branch?: string;
}
