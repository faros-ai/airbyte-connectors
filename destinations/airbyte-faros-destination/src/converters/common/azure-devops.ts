import {User} from 'faros-airbyte-common/azure-devops';

// TODO: Make it return OrgKey
export function getOrganizationFromUrl(url: string): string | undefined {
  try {
    const parsed = new URL(url);
    const parts = parsed.pathname.split('/');

    if (parts.length < 2 || parts[1] === '') {
      return undefined;
    }

    return parts[1];
  } catch (error) {
    return undefined;
  }
}

export function getProjectFromUrl(url: string): string | undefined {
  try {
    const parsed = new URL(url);
    const parts = parsed.pathname.split('/');

    if (parts.length < 3 || parts[2] === '') {
      return undefined;
    }

    return parts[2];
  } catch (error) {
    return undefined;
  }
}

// Support both principalName and uniqueName for AzureDevOps Server
export function getUniqueName(userItem: User): string | undefined {
  if ('principalName' in userItem && Boolean(userItem.principalName)) {
    return userItem.principalName;
  }
  if ('uniqueName' in userItem && Boolean(userItem.uniqueName)) {
    return userItem.uniqueName;
  }
  return undefined;
}
