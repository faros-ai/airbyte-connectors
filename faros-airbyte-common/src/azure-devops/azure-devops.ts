import {AxiosInstance} from 'axios';
import {getPersonalAccessTokenHandler, WebApi} from 'azure-devops-node-api';
import {
  IdentityRef,
  TeamMember,
} from 'azure-devops-node-api/interfaces/common/VSSInterfaces';
import {
  TeamProject,
  TeamProjectReference,
  WebApiTeam,
} from 'azure-devops-node-api/interfaces/CoreInterfaces';
import {GraphUser} from 'azure-devops-node-api/interfaces/GraphInterfaces';
import {AirbyteLogger, base64Encode} from 'faros-airbyte-cdk';
import {makeAxiosInstanceWithRetry} from 'faros-js-client';
import {Memoize} from 'typescript-memoize';
import VError from 'verror';

import * as types from './types';

interface ResponseWithContinuationToken {
  continuationToken?: string;
}

type ItemFromResponseWithContinuationToken<
  T extends ResponseWithContinuationToken,
  K extends keyof T,
> = T[K] extends (infer U)[] ? U : never;

const CLOUD_API_URL = 'https://dev.azure.com';
const DEFAULT_GRAPH_API_URL = 'https://vssps.dev.azure.com';
const DEFAULT_GRAPH_API_VERSION = '7.1-preview.1';
const DEFAULT_REQUEST_TIMEOUT = 300_000;
const DEFAULT_MAX_RETRIES = 3;
const DEFAULT_PAGE_SIZE = 100;
const DEFAULT_CUTOFF_DAYS = 90;

/**
 * This abstract class is used to create Azure DevOps clients.
 * Extended by Azure DevOps sources that need to access Azure DevOps APIs,
 * such as Azure Repos, Azure Pipelines, Azure Work Items, etc.
 *
 * @abstract
 * @class AzureDevOps
 */
export abstract class AzureDevOps {
  private static azureDevOps: AzureDevOps = null;

  constructor(
    protected readonly client: types.AzureDevOpsClient,
    protected readonly instanceType: 'cloud' | 'server',
    protected readonly cutoffDays: number = DEFAULT_CUTOFF_DAYS,
    protected readonly top: number = DEFAULT_PAGE_SIZE,
    protected readonly logger: AirbyteLogger
  ) {}

  static async instance<T extends AzureDevOps>(
    this: new (...args: any[]) => T,
    config: types.AzureDevOpsConfig,
    logger: AirbyteLogger,
    ...additionalArgs: any[]
  ): Promise<T> {
    if (AzureDevOps.azureDevOps) {
      return AzureDevOps.azureDevOps as T;
    }

    AzureDevOps.validateAuth(config);

    if (config.instance?.type) {
      logger.info(`Azure DevOps instance type: ${config.instance.type}`);
    } else {
      logger.info(
        'No Azure DevOps instance type provided, defaulting to ' +
          'Azure DevOps Services (cloud)'
      );
    }

    const baseUrl = AzureDevOps.getBaseUrl(
      config?.instance,
      config.organization
    );
    const timeout = config.request_timeout ?? DEFAULT_REQUEST_TIMEOUT;
    const maxRetries = config.max_retries ?? DEFAULT_MAX_RETRIES;
    const webApi = await AzureDevOps.createWebAPI(
      baseUrl,
      config.access_token,
      timeout,
      maxRetries
    );
    const restApi = AzureDevOps.createRestAPI(
      baseUrl,
      config.access_token,
      timeout,
      maxRetries,
      logger
    );

    const graphApi = AzureDevOps.createRestAPI(
      `${DEFAULT_GRAPH_API_URL}/${config.organization}/_apis/graph`,
      config.access_token,
      timeout,
      maxRetries,
      logger
    );

    const client = {
      build: await webApi.getBuildApi(),
      core: await webApi.getCoreApi(),
      wit: await webApi.getWorkItemTrackingApi(),
      git: await webApi.getGitApi(),
      pipelines: await webApi.getPipelinesApi(),
      release: await webApi.getReleaseApi(),
      test: await webApi.getTestApi(),
      rest: restApi,
      graph: graphApi,
    };

    const pageSize = isNaN(config.page_size) ? undefined : config.page_size;
    if (pageSize && pageSize <= 1) {
      logger.warn(
        'Page size must be greater than 1, will use default minimum of 2'
      );
    }

    const top = Math.max(2, pageSize ?? DEFAULT_PAGE_SIZE);
    const cutoffDays = config.cutoff_days ?? DEFAULT_CUTOFF_DAYS;

    AzureDevOps.azureDevOps = new this(
      client,
      config.instance?.type,
      cutoffDays,
      top,
      logger,
      ...additionalArgs
    );
    return AzureDevOps.azureDevOps as T;
  }

  static getBaseUrl(
    instance: types.AzureDevOpsInstance,
    organization: string
  ): string {
    if (instance?.type?.toLowerCase() !== 'server') {
      return `${CLOUD_API_URL}/${organization}`;
    }

    // Type assertion since we've verified it's a server instance
    const serverInstance = instance as types.DevOpsServer;
    const apiUrl = serverInstance.api_url?.trim();
    if (!apiUrl) {
      throw new VError(
        'api_url must not be an empty string for server instance type'
      );
    }

    try {
      new URL(apiUrl);
    } catch (error) {
      throw new VError(`Invalid URL: ${serverInstance.api_url}`);
    }
    return `${apiUrl}/${organization}`;
  }

  static validateAuth(config: types.AzureDevOpsConfig): void {
    if (!config.access_token) {
      throw new VError('access_token must not be an empty string');
    }
    if (!config.organization) {
      throw new VError('organization must not be an empty string');
    }
  }

  // Create Azure DevOps Node API clients
  static async createWebAPI(
    baseUrl: string,
    accessToken: string,
    timeout: number = DEFAULT_REQUEST_TIMEOUT,
    maxRetries: number = DEFAULT_MAX_RETRIES
  ): Promise<WebApi> {
    const authHandler = getPersonalAccessTokenHandler(accessToken);
    return new WebApi(baseUrl, authHandler, {
      socketTimeout: timeout,
      allowRetries: true,
      maxRetries,
      globalAgentOptions: {
        keepAlive: true,
        timeout,
      },
    });
  }

  static createRestAPI(
    baseUrl: string,
    access_token: string,
    timeout: number,
    maxRetries: number,
    logger: AirbyteLogger
  ): AxiosInstance | undefined {
    const base64EncodedToken = base64Encode(`:${access_token}`);
    return makeAxiosInstanceWithRetry(
      {
        baseURL: baseUrl,
        timeout,
        maxContentLength: Infinity,
        maxBodyLength: Infinity,
        headers: {
          Authorization: `Basic ${base64EncodedToken}`,
        },
      },
      logger.asPino(),
      maxRetries,
      1000 // Default retry delay
    );
  }

  // Overload for standard skip-based pagination (no options)
  protected getPaginated<T>(
    getFn: (top: number, skipOrToken: number | string) => Promise<Array<T>>
  ): AsyncGenerator<T>;

  // Overload for continuation token pagination workaround
  protected getPaginated<T>(
    getFn: (top: number, skipOrToken: number | string) => Promise<Array<T>>,
    options: {
      useContinuationToken: true;
      continuationTokenParam: string;
    }
  ): AsyncGenerator<T>;

  // Overload for continuation token pagination
  protected getPaginated<
    T extends ResponseWithContinuationToken,
    K extends keyof T,
  >(
    getFn: (top: number, skipOrToken: number | string) => Promise<T>,
    options: {
      useContinuationToken: true;
      itemsField: K;
    }
  ): AsyncGenerator<ItemFromResponseWithContinuationToken<T, K>>;

  // Implementation signature
  protected async *getPaginated<T, K extends keyof T = any>(
    getFn: (top: number, skipOrToken: number | string) => Promise<Array<T> | T>,
    options?: {
      useContinuationToken?: boolean;
      continuationTokenParam?: string;
      itemsField?: string;
    }
  ): AsyncGenerator<T | ItemFromResponseWithContinuationToken<T, K>> {
    if (options?.useContinuationToken) {
      if (!options?.itemsField) {
        yield* this.paginateWithContinuationTokenWorkaround(
          getFn as (
            top: number,
            continuationToken: number | string
          ) => Promise<Array<T>>,
          options.continuationTokenParam
        );
      } else {
        yield* this.paginateWithContinuationToken(
          getFn as (
            top: number,
            continuationToken: number | string
          ) => Promise<T>,
          options.itemsField
        );
      }
    } else {
      // Skip-based pagination
      yield* this.paginateWithSkip(
        getFn as (top: number, skip: number) => Promise<Array<T>>
      );
    }
  }

  private async *paginateWithSkip<T>(
    getFn: (top: number, skip: number) => Promise<Array<T>>
  ): AsyncGenerator<T> {
    let resCount = 0;
    let skip = 0;
    const top = this.top;

    do {
      const result = await getFn(top, skip);
      if (result?.length) yield* result;
      resCount = result?.length ?? 0;
      skip += resCount;
    } while (resCount >= top);
  }

  // Workaround for Azure DevOps API pagination not returning the continuation token
  // https://github.com/microsoft/azure-devops-node-api/issues/609
  private async *paginateWithContinuationTokenWorkaround<T>(
    getFn: (
      top: number,
      continuationToken: string | number
    ) => Promise<Array<T>>,
    continuationTokenParam: string
  ): AsyncGenerator<T> {
    const top = this.top;
    let hasNext = false;
    let pages = 0;
    let continuationToken = undefined;

    do {
      const result = await getFn(top, continuationToken);
      if (!result) {
        this.logger.warn(
          'Failed to fetch results, received empty result. Skipping...'
        );
        return;
      }

      if (!Array.isArray(result)) {
        throw new VError(
          `Expected array result but got ${typeof result} ${JSON.stringify(
            result
          )}`
        );
      }

      // Remove the first record when we have a continuation token to avoid duplication
      // since we already processed it in the previous page
      const records = continuationToken ? result.slice(1) : result;
      yield* records;

      continuationToken =
        result.length === top
          ? result.at(-1)[continuationTokenParam]
          : undefined;

      hasNext = Boolean(continuationToken);
      pages++;
      this.logger.debug(
        hasNext
          ? `Fetching next page using continuation token ${continuationToken}`
          : `Finished fetching ${pages} pages`
      );
    } while (hasNext);
  }

  // Handle responses with nested arrays and response-level continuation tokens
  private async *paginateWithContinuationToken<
    T extends ResponseWithContinuationToken,
    K extends keyof T,
  >(
    getFn: (top: number, continuationToken: string | number) => Promise<T>,
    itemsField: string
  ): AsyncGenerator<ItemFromResponseWithContinuationToken<T, K>> {
    const top = this.top;
    let hasNext = false;
    let pages = 0;
    let continuationToken = undefined;

    do {
      const result = await getFn(top, continuationToken);
      if (!result) {
        this.logger.warn(
          'Failed to fetch results, received empty result. Skipping...'
        );
        return;
      }

      // Extract the items array from the nested response
      const items = result[itemsField] as Array<
        ItemFromResponseWithContinuationToken<T, K>
      >;
      if (!Array.isArray(items)) {
        throw new VError(
          `Expected array at ${itemsField} but got ${typeof items} ${JSON.stringify(
            items
          )}`
        );
      }

      yield* items;

      // Get continuation token from the response level
      continuationToken = result.continuationToken;
      hasNext = Boolean(continuationToken);
      pages++;

      this.logger.debug(
        hasNext
          ? `Fetched page ${pages} with ${items.length} items, continuing with token ${continuationToken}`
          : `Finished fetching ${pages} pages`
      );
    } while (hasNext);
  }

  @Memoize()
  async getProjects(
    projects?: ReadonlyArray<string>
  ): Promise<ReadonlyArray<TeamProject>> {
    if (!projects?.length) {
      return await this.getAllProjects();
    }

    const allProjects = [];
    for (const project of projects) {
      const res = await this.client.core.getProject(project);
      if (res) {
        allProjects.push(res);
      } else {
        this.logger.warn(`Project ${project} in config not found. Skipping`);
      }
    }
    return allProjects;
  }

  private async getAllProjects(): Promise<ReadonlyArray<TeamProjectReference>> {
    const projects = [];
    const getProjectsFn = async (
      top: number,
      skip: number | string
    ): Promise<Array<TeamProjectReference>> => {
      const res = await this.client.core.getProjects(
        'wellFormed',
        top,
        skip as number
      );
      return Array.from(res.values());
    };

    for await (const project of this.getPaginated(getProjectsFn)) {
      projects.push(project);
    }
    return projects;
  }

  protected async *getTeams(projectId: string): AsyncGenerator<WebApiTeam> {
    const getTeamsFn = (
      pageSize: number,
      skip: number | string
    ): Promise<WebApiTeam[]> =>
      this.client.core.getTeams(projectId, false, pageSize, skip as number);

    yield* this.getPaginated<WebApiTeam>(getTeamsFn);
  }

  protected async *getTeamMembers(
    projectId: string,
    teamId: string
  ): AsyncGenerator<IdentityRef> {
    const getMembersFn = (
      top: number,
      skip: number | string
    ): Promise<TeamMember[]> =>
      this.client.core.getTeamMembersWithExtendedProperties(
        projectId,
        teamId,
        top,
        skip as number
      );

    for await (const member of this.getPaginated(getMembersFn)) {
      yield member.identity;
    }
  }

  // For Azure DevOps Services (cloud) instance, use the graph API to fetch users.
  // For Azure DevOps Server instance, fetch users through teams and team members.
  async *getUsers(
    projects?: ReadonlyArray<string>
  ): AsyncGenerator<types.User> {
    if (this.instanceType?.toLowerCase() !== 'server') {
      yield* this.getGraphUsers();
      return;
    }
    yield* this.getTeamUsers(projects);
  }

  private async *getGraphUsers(): AsyncGenerator<GraphUser> {
    let continuationToken: string;
    do {
      const res = await this.client.graph.get<types.GraphUserResponse>(
        `users`,
        {
          params: {
            'api-version': DEFAULT_GRAPH_API_VERSION,
            subjectTypes: 'msa,aad,imp',
            continuationToken,
          },
        }
      );
      continuationToken = res?.headers?.['X-MS-ContinuationToken'];
      for (const item of res?.data?.value ?? []) {
        yield item;
      }
    } while (continuationToken);
  }

  private async *getTeamUsers(
    projects?: ReadonlyArray<string>
  ): AsyncGenerator<IdentityRef> {
    const seenUsers = new Set<string>();
    let teams = 0;

    for (const project of await this.getProjects(projects)) {
      for await (const team of this.getTeams(project.id)) {
        teams++;
        for await (const member of this.getTeamMembers(project.id, team.id)) {
          if (!seenUsers.has(member.uniqueName)) {
            seenUsers.add(member.uniqueName);
            yield member;
          }
        }
      }
    }
    this.logger.debug(`Fetched ${seenUsers.size} members from ${teams} teams`);
  }
}
