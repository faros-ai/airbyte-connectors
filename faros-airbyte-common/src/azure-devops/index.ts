export * from './types';
export {AzureDevOps} from './azure-devops';
export * from './utils';
