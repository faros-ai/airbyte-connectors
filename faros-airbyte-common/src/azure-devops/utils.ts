import {User} from './types';

/**
 * Gets a unique identifier for an Azure DevOps user.
 * Tries principalName first, then falls back to uniqueName.
 * This handles differences between Azure DevOps Cloud and Server.
 *
 * @param user - The Azure DevOps user object
 * @returns The user's unique identifier, or undefined if neither field exists
 */
export function getUserIdentifier(user: User): string | undefined {
  if ('principalName' in user && user.principalName) {
    return user.principalName;
  }
  if ('uniqueName' in user && user.uniqueName) {
    return user.uniqueName;
  }
  return undefined;
}
