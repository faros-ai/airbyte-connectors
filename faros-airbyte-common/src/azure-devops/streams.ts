import {TeamProject} from 'azure-devops-node-api/interfaces/CoreInterfaces';
import {AirbyteLogger, AirbyteStreamBase, StreamKey} from 'faros-airbyte-cdk';

import {AzureDevOps} from './azure-devops';
import {AzureDevOpsConfig} from './types';

export type ProjectStreamSlice = {
  name: string;
  id: string;
};

export abstract class AzureDevOpsStreamBase extends AirbyteStreamBase {
  constructor(
    protected readonly config: AzureDevOpsConfig,
    protected readonly logger: AirbyteLogger
  ) {
    super(logger);
  }

  get primaryKey(): StreamKey {
    return 'id';
  }
}

export abstract class StreamWithProjectSlices extends AzureDevOpsStreamBase {
  async *streamSlices(): AsyncGenerator<TeamProject> {
    const azureDevOps = await AzureDevOps.instance(this.config, this.logger);
    const projects = await azureDevOps.getProjects(this.config.projects);
    yield* projects;
  }
}
