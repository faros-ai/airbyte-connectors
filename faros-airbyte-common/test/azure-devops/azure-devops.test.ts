import {
  AirbyteLogLevel,
  AirbyteSourceLogger,
  readTestResourceAsJSON,
} from 'faros-airbyte-cdk';

import {AzureDevOps} from '../../src/azure-devops/azure-devops';
import {
  AzureDevOpsClient,
  AzureDevOpsConfig,
} from '../../src/azure-devops/types';

jest.mock('azure-devops-node-api');

const paginationItems = [
  {id: 1, name: 'item1'},
  {id: 2, name: 'item2'},
  {id: 3, name: 'item3'},
  {id: 4, name: 'item4'},
  {id: 5, name: 'item5'},
];

describe('client', () => {
  class TestAzureDevOps extends AzureDevOps {}

  const logger = new AirbyteSourceLogger(
    // Shush messages in tests, unless in debug
    process.env.LOG_LEVEL === 'debug'
      ? AirbyteLogLevel.DEBUG
      : AirbyteLogLevel.INFO
  );

  const config = {
    access_token: 'access_token',
    organization: 'organization',
    instance: {
      type: 'cloud',
    },
  } as AzureDevOpsConfig;

  beforeEach(() => {
    (AzureDevOps as any).azureDevOps = null;
  });

  test('valid config for azure devops services (cloud)', async () => {
    const azureDevOps = await TestAzureDevOps.instance(config, logger);
    expect(azureDevOps).toBeInstanceOf(AzureDevOps);
  });

  test('valid config for azure devops server', async () => {
    const config = {
      access_token: 'access_token',
      organization: 'organization',
      instance: {
        type: 'server',
        api_url: 'https://my-server.com',
      },
    } as AzureDevOpsConfig;

    const azureDevOps = await TestAzureDevOps.instance(config, logger);
    expect(azureDevOps).toBeInstanceOf(AzureDevOps);
  });

  test('missing access_token in config', async () => {
    const config = {
      access_token: null,
      organization: 'organization',
    };
    await expect(TestAzureDevOps.instance(config, logger)).rejects.toThrow(
      'access_token must not be an empty string'
    );
  });

  test('missing organization in config', async () => {
    const config = {
      access_token: 'access_token',
      organization: null,
    };
    await expect(TestAzureDevOps.instance(config, logger)).rejects.toThrow(
      'organization must not be an empty string'
    );
  });

  test('missing api url for server instance', async () => {
    const config = {
      access_token: 'access_token',
      organization: 'organization',
      instance: {
        type: 'server',
        api_url: null,
      },
    } as AzureDevOpsConfig;
    await expect(TestAzureDevOps.instance(config, logger)).rejects.toThrow(
      'api_url must not be an empty string for server instance type'
    );
  });

  test('invalid api url for server instance', async () => {
    const config = {
      access_token: 'access_token',
      organization: 'organization',
      instance: {
        type: 'server',
        api_url: 'invalid_api_url',
      },
    } as AzureDevOpsConfig;
    await expect(TestAzureDevOps.instance(config, logger)).rejects.toThrow(
      'Invalid URL: invalid_api_url'
    );
  });

  test('paginate with skip parameter', async () => {
    const getPage = (pageSize: number, skip: number) => {
      return paginationItems.slice(skip, skip + pageSize);
    };

    for (const pageSize of [2, 3, 4]) {
      const azureDevOps = await TestAzureDevOps.instance(
        {...config, page_size: pageSize},
        logger
      );

      const mockGetFn = jest.fn((top: number, skip: number | string) =>
        Promise.resolve(getPage(top, skip as number))
      );

      const results = [];
      for await (const item of azureDevOps['getPaginated'](mockGetFn)) {
        results.push(item);
      }

      // Verify all items are returned
      expect(results).toEqual(paginationItems);

      // Verify calls
      const expectedCalls = Math.ceil(paginationItems.length / pageSize);
      expect(mockGetFn).toHaveBeenCalledTimes(expectedCalls);

      // Verify parameters for each call
      const calls = mockGetFn.mock.calls;
      calls.forEach((call, index) => {
        const top = Math.max(2, pageSize);
        expect(call[0]).toEqual(top); // top parameter
        expect(call[1]).toEqual(index * top); // skip parameter
      });

      mockGetFn.mockClear();
      (AzureDevOps as any).azureDevOps = null;
    }
  });

  test('paginate with continuation token', async () => {
    const getPage = (pageSize: number, continuationToken?: number) => {
      const startIndex = continuationToken
        ? paginationItems.findIndex((item) => item.id === continuationToken)
        : 0;
      return paginationItems.slice(startIndex, startIndex + pageSize);
    };

    const pageSizeCalls = [
      [1, 5],
      [2, 5],
      [5, 2],
      [10, 1],
    ];

    for (const [pageSize, numCalls] of pageSizeCalls) {
      const azureDevOps = await TestAzureDevOps.instance(
        {...config, page_size: pageSize},
        logger
      );

      const mockGetFn = jest.fn(
        (top: number, continuationToken?: string | number) =>
          Promise.resolve(getPage(top, continuationToken as number))
      );

      const results = [];
      for await (const item of azureDevOps['getPaginated'](mockGetFn, {
        useContinuationToken: true,
        continuationTokenParam: 'id',
      })) {
        results.push(item);
      }

      expect(results).toEqual(paginationItems);
      expect(mockGetFn).toHaveBeenCalledTimes(numCalls);

      const calls = mockGetFn.mock.calls;
      const effectivePageSize = Math.max(2, pageSize);
      calls.forEach((call, index) => {
        expect(call[0]).toEqual(effectivePageSize);

        if (index === 0) {
          expect(call[1]).toBeUndefined();
        } else {
          const lastItemIndex = index * (effectivePageSize - 1);
          expect(call[1]).toEqual(paginationItems.at(lastItemIndex).id);
        }
      });

      mockGetFn.mockClear();
      (AzureDevOps as any).azureDevOps = null;
    }
  });

  test('paginate with continuation token handles non-array results', async () => {
    const azureDevOps = await TestAzureDevOps.instance(config, logger);
    const mockGetFn = jest.fn().mockResolvedValue({error: 'invalid response'});

    await expect(async () => {
      for await (const _ of azureDevOps['getPaginated'](mockGetFn, {
        useContinuationToken: true,
        continuationTokenParam: 'id',
      })) {
        // Do nothing
      }
    }).rejects.toThrow('Expected array result but got object');
  });

  test('paginate with continuation token handles empty results', async () => {
    const azureDevOps = await TestAzureDevOps.instance(config, logger);
    const mockGetFn = jest.fn().mockResolvedValue(undefined);

    const results = [];
    for await (const item of azureDevOps['getPaginated'](mockGetFn, {
      useContinuationToken: true,
      continuationTokenParam: 'id',
    })) {
      results.push(item);
    }
    expect(mockGetFn).toHaveBeenCalledTimes(1);
    expect(results).toEqual([]);
  });

  test('get projects (no projects provided)', async () => {
    TestAzureDevOps.instance = jest.fn().mockImplementation(() => {
      return new TestAzureDevOps(
        {
          core: {
            getProjects: jest
              .fn()
              .mockResolvedValueOnce(readTestResourceAsJSON('projects.json')),
          },
        } as unknown as AzureDevOpsClient,
        90,
        100,
        logger
      );
    });

    const azureDevOps = await TestAzureDevOps.instance(config, logger);
    expect(await azureDevOps.getProjects()).toMatchSnapshot();
  });

  test('get projects (empty list provided)', async () => {
    TestAzureDevOps.instance = jest.fn().mockImplementation(() => {
      return new TestAzureDevOps(
        {
          core: {
            getProjects: jest
              .fn()
              .mockResolvedValueOnce(readTestResourceAsJSON('projects.json')),
          },
        } as unknown as AzureDevOpsClient,
        90,
        100,
        logger
      );
    });

    const azureDevOps = await TestAzureDevOps.instance(config, logger);
    expect(await azureDevOps.getProjects([])).toMatchSnapshot();
  });

  test('get projects (list provided)', async () => {
    TestAzureDevOps.instance = jest.fn().mockImplementation(() => {
      const projects = readTestResourceAsJSON('projects.json');
      return new TestAzureDevOps(
        {
          core: {
            getProject: jest
              .fn()
              .mockResolvedValueOnce(projects.at(2))
              .mockResolvedValueOnce(projects.at(1)),
          },
        } as unknown as AzureDevOpsClient,
        90,
        100,
        logger
      );
    });

    const azureDevOps = await TestAzureDevOps.instance(config, logger);
    expect(
      await azureDevOps.getProjects(['TestGit', 'Fabrikam-Fiber-Git'])
    ).toMatchSnapshot();
  });

  test('get projects (invalid project provided)', async () => {
    TestAzureDevOps.instance = jest.fn().mockImplementation(() => {
      const projects = readTestResourceAsJSON('projects.json');
      return new TestAzureDevOps(
        {
          core: {
            getProject: jest
              .fn()
              .mockResolvedValueOnce(projects.at(2))
              .mockResolvedValueOnce(undefined),
          },
        } as unknown as AzureDevOpsClient,
        90,
        100,
        logger
      );
    });

    const azureDevOps = await TestAzureDevOps.instance(config, logger);
    expect(
      await azureDevOps.getProjects(['TestGit', 'InvalidProject'])
    ).toMatchSnapshot();
  });
});
