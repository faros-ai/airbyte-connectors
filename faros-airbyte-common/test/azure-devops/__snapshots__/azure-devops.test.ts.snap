// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client get projects (empty list provided) 1`] = `
Array [
  Object {
    "description": "Team Foundation Version Control projects.",
    "id": "eb6e4656-77fc-42a1-9181-4c6d8e9da5d1",
    "name": "Fabrikam-Fiber-TFVC",
    "state": "wellFormed",
    "url": "https://fabrikam:8080/_apis/projects/eb6e4656-77fc-42a1-9181-4c6d8e9da5d1",
  },
  Object {
    "description": "Git projects",
    "id": "6ce954b1-ce1f-45d1-b94d-e6bf2464ba2c",
    "name": "Fabrikam-Fiber-Git",
    "state": "wellFormed",
    "url": "https://fabrikam:8080/_apis/projects/6ce954b1-ce1f-45d1-b94d-e6bf2464ba2c",
  },
  Object {
    "id": "281f9a5b-af0d-49b4-a1df-fe6f5e5f84d0",
    "name": "TestGit",
    "state": "wellFormed",
    "url": "https://fabrikam:8080/_apis/projects/281f9a5b-af0d-49b4-a1df-fe6f5e5f84d0",
  },
]
`;

exports[`client get projects (invalid project provided) 1`] = `
Array [
  Object {
    "id": "281f9a5b-af0d-49b4-a1df-fe6f5e5f84d0",
    "name": "TestGit",
    "state": "wellFormed",
    "url": "https://fabrikam:8080/_apis/projects/281f9a5b-af0d-49b4-a1df-fe6f5e5f84d0",
  },
]
`;

exports[`client get projects (list provided) 1`] = `
Array [
  Object {
    "id": "281f9a5b-af0d-49b4-a1df-fe6f5e5f84d0",
    "name": "TestGit",
    "state": "wellFormed",
    "url": "https://fabrikam:8080/_apis/projects/281f9a5b-af0d-49b4-a1df-fe6f5e5f84d0",
  },
  Object {
    "description": "Git projects",
    "id": "6ce954b1-ce1f-45d1-b94d-e6bf2464ba2c",
    "name": "Fabrikam-Fiber-Git",
    "state": "wellFormed",
    "url": "https://fabrikam:8080/_apis/projects/6ce954b1-ce1f-45d1-b94d-e6bf2464ba2c",
  },
]
`;

exports[`client get projects (no projects provided) 1`] = `
Array [
  Object {
    "description": "Team Foundation Version Control projects.",
    "id": "eb6e4656-77fc-42a1-9181-4c6d8e9da5d1",
    "name": "Fabrikam-Fiber-TFVC",
    "state": "wellFormed",
    "url": "https://fabrikam:8080/_apis/projects/eb6e4656-77fc-42a1-9181-4c6d8e9da5d1",
  },
  Object {
    "description": "Git projects",
    "id": "6ce954b1-ce1f-45d1-b94d-e6bf2464ba2c",
    "name": "Fabrikam-Fiber-Git",
    "state": "wellFormed",
    "url": "https://fabrikam:8080/_apis/projects/6ce954b1-ce1f-45d1-b94d-e6bf2464ba2c",
  },
  Object {
    "id": "281f9a5b-af0d-49b4-a1df-fe6f5e5f84d0",
    "name": "TestGit",
    "state": "wellFormed",
    "url": "https://fabrikam:8080/_apis/projects/281f9a5b-af0d-49b4-a1df-fe6f5e5f84d0",
  },
]
`;
